### PR TITLE
Compute profiling/shared based on library deps only.

### LINKED ISSUE #3789: Don't build setup/executable dependencies with profiling, if profiling library needed
It's pointless and a waste of time.

Steps to reproduce: `--enable-profiling` on a package with a dependency which has a Custom setup that requires an old version of Cabal.

### PR DESCRIPTION
**Pull request by ezyang**
Previously we'd build Cabal and custom setup dependencies
with profiling even though it wasn't necessary. Using a
refined dependency graph fixes the problem.

If looking at a Graph with different set of neighbors becomes
a common op, maybe we should add some functions to handle
this specific case.

Fixes #3789.

### CODE REVIEW

File: cabal-install/Distribution/Client/ProjectPlanning.hs [hunk 1]
```diff
@@ -1573,6 +1578,18 @@ elaborateInstallPlan platform compiler compilerprogdb
       -- + vanilla libs & exes, exe needs lib, recursive
       -- + ghci or shared lib needed by TH, recursive, ghc version dependent
 
+newtype LibDepSolverPlanPackage
```
Comment (23Skidoo) [inline cabal-install/Distribution/Client/ProjectPlanning.hs]: A Haddock comment here would be nice.

File: cabal-install/Distribution/Client/SolverInstallPlan.hs [hunk 2]
```diff
@@ -143,6 +144,9 @@ new indepGoals index =
 toList :: SolverInstallPlan -> [SolverPlanPackage]
 toList = Graph.toList . planIndex
 
+toMap :: SolverInstallPlan -> Map SolverId SolverPlanPackage
```
Comment (23Skidoo) [inline cabal-install/Distribution/Client/SolverInstallPlan.hs]: You don't actually use it, though.

**Comment by 23Skidoo** [comment]
LGTM.

File: cabal-install/Distribution/Client/SolverInstallPlan.hs (hunk 2 above)
Comment (ezyang) [inline cabal-install/Distribution/Client/SolverInstallPlan.hs]: Yeah, I thought I needed it, turned out I didn't, but it seemed generally useful so I didn't remove it.

**Comment by dcoutts** [comment]
LGTM

**Comment by 23Skidoo** [comment]
Merged, thanks!

**Pull request merged**